Zend/zend_API.c

```diff
@@ -1846,6 +1846,39 @@ ZEND_API zend_result object_init_ex(zval *arg, zend_class_entry *class_type) /*
 }
 /* }}} */
 
+ZEND_API zend_result object_init_with_constructor(zval *arg, zend_class_entry *class_type, uint32_t param_count, zval *params, HashTable *named_params) /* {{{ */
+{
+	zend_result status = _object_and_properties_init(arg, class_type, NULL);
+	if (UNEXPECTED(status == FAILURE)) {
+		return FAILURE;
+	}
+	zend_object *obj = Z_OBJ_P(arg);
+	zend_function *constructor = obj->handlers->get_constructor(obj);
+	if (UNEXPECTED(constructor == NULL)) {
+		return FAILURE;
+	}
+	/* A constructor should not return a value, however if an exception is thrown
+	 * zend_call_known_function() will set the retval to IS_UNDEF */
+	zval retval;
+	zend_call_known_function(
+		constructor,
+		obj,
+		class_type,
+		/* retval */ &retval,
+		param_count,
+		params,
+		named_params
+	);
+	if (Z_TYPE(retval) == IS_UNDEF) {
+		return FAILURE;
```

Comment:
TIL about `zend_object_store_ctor_failed`, I think it indeed makes sense to call it.

I was thinking I had a bug in not releasing the object, but I couldn't produce it with a test so I left it, will try and see if I can come up with something and add some destructors to the tests.

Reply:
The object is useless because it's incompletely initialized, so the caller can't do much with it. I think I'd prefer it it were released and not returned at all.

Reply:
This discussion reminded me of my PR #14161 (“Add zend_get_attribute_object()”), which also automatically destructs the object on constructor failure.

For the `#[\Deprecated]` PR, I've made a follow-up to also `ZVAL_UNDEF()` the zval then, making it safe to unconditionally `zval_ptr_dtor()` the zval no matter if the constructor failed or not:

https://github.com/php/php-src/pull/11293/commits/4211f4dbdd0c4245a0ea2353ff711aad55ee4b12

I would suggest the same here: If the constructor fails, the object is destroyed and the zval UNDEF'd.

Reply:
Yeah this is what I am doing